components/core/tests/test-NetworkReader.cpp

```diff
@@ -188,3 +208,47 @@ TEST_CASE("network_reader_illegal_offset", "[NetworkReader]") {
     size_t pos{};
     REQUIRE((clp::ErrorCode_Failure == reader.try_get_pos(pos)));
 }
+
+TEST_CASE("network_reader_with_http_header_kv_pairs", "[NetworkReader]") {
+    std::unordered_map<std::string, std::string> regular_http_header_kv_pairs;
+    // We use httpbin (https://httpbin.org/) to test the custom headers. This request will return a
+    // JSON object that contains the custom headers. We check if the headers are in the response.
+    constexpr int cNumRegularTestHeaders{10};
+    for (size_t i{0}; i < cNumRegularTestHeaders; i++) {
+        regular_http_header_kv_pairs.emplace(
+                fmt::format("Unit-Test-Key{}", i),
+                fmt::format("Unit-Test-Value{}", i)
+        );
+    }
+    clp::NetworkReader regular_reader{
+            "https://httpbin.org/headers",
+            0,
+            false,
+            clp::CurlDownloadHandler::cDefaultOverallTimeout,
+            clp::CurlDownloadHandler::cDefaultConnectionTimeout,
+            clp::NetworkReader::cDefaultBufferPoolSize,
+            clp::NetworkReader::cDefaultBufferSize,
+            regular_http_header_kv_pairs
+    };
+    auto content{nlohmann::json::parse(get_content(regular_reader))};
+    if (content.is_array() && 1 == content.size()) {
+        content = content.at(0);
+    }
+    auto const& headers{content.at("headers")};
+    REQUIRE(assert_curl_error_code(CURLE_OK, regular_reader));
+    for (int i = 0; i < cNumRegularTestHeaders; i++) {
+        REQUIRE((
+                fmt::format("Unit-Test-Value{}", i) == headers.at(fmt::format("Unit-Test-Key{}", i))
+        ));
+    }
+    // The following headers are determined by offset and disable_cache, which should not be
+    // overrided by custom headers.
+    REQUIRE(test_illegal_header("Range", "bytes=100-"));
+    REQUIRE(test_illegal_header("Cache-Control", "no-cache"));
+    REQUIRE(test_illegal_header("Pragma", "no-cache"));
+    // The following headers contain illegal header values, the requests should be rejected.
+    REQUIRE(test_illegal_header("Legal-Name1", "CRLF\n"));
+    REQUIRE(test_illegal_header("Legal-Name2", "CRLF\r"));
+    REQUIRE(test_illegal_header("Legal-Name3", "CRLF\n\r"));
+    REQUIRE(test_illegal_header("Legal-Name4", "CRLF\r\n"));
```

Comment:
as said above, let's move this into a separate test-case with `GENERATE` to replace `test_illegal_header`